### PR TITLE
α-GAN (AlphaGAN)

### PR DESCRIPTION
**Pull request by martyn**
https://arxiv.org/pdf/1706.04987.pdf

Adds the command line flag `--autoencode alphagan`

* Walking a network (still training)
random_walk

Sampled with `--sampler autoencode`

### CODE REVIEW

File: hypergan/configurations/improved-gan-resizable.json [hunk 1]
```diff
@@ -30,8 +29,7 @@
   {
     "class": "function:hypergan.losses.standard_loss.StandardLoss",
     "reduce": "reduce_mean",
-    "improved": true,
-    "minibatch": true
+    "improved": true
```
Comment (martyn) [inline hypergan/configurations/improved-gan-resizable.json]: Minibatch doesn't work here.  Maybe we can output an error if `gan.batch_size()*2 != shape(net)[0]`. 

The reason is that we stack 3 values on the Discriminator instead of 2(X, reconstruction of X, and a random generated sample).

File: hypergan/discriminators/fully_connected_discriminator.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class FullyConnectedDiscriminator(BaseDiscriminator):
+    def build(self, net):
+        config = self.config
+        gan = self.gan
+        ops = self.ops
+        activation = ops.lookup(config.activation or 'lrelu')
+        final_activation = ops.lookup(config.final_activation or 'tanh')
+
+        net = ops.linear(net, 512)
```
Comment (martyn) [inline hypergan/discriminators/fully_connected_discriminator.py]: Maybe we can parameterize this.

File: hypergan/losses/base_loss.py [hunk 3]
```diff
@@ -45,6 +53,8 @@ def create(self):
         self.metrics = self.metrics or sample_metrics
 
         self.sample = [d_loss, g_loss]
+        self.d_loss = d_loss
```
Comment (martyn) [inline hypergan/losses/base_loss.py]: These are probably better variable names for Loss instance methods.  I think the .sample pattern starts to fall apart for loss terms(we already exclude access to self.ops).

File: hypergan/trainers/alpha_trainer.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        gan = self.gan
+        sess = gan.session
+        config = self.config
+        losses = self.losses
+
+        for i, _ in enumerate(losses):
+            loss = losses[i]
+            optimizer = self.optimizers[i]
+            #metrics = loss.metrics
+            _ = sess.run(optimizer)
```
Comment (martyn) [inline hypergan/trainers/alpha_trainer.py]: Doesn't output metrics(one more reason `-c began-optimized` fails)

File: hypergan/trainers/base_trainer.py [hunk 5]
```diff
@@ -32,3 +34,43 @@ def step(self, feed_dict={}):
...
+                return None
+            return (tf.clip_by_value(grad, -cap, cap), var)
+
+        capped_gvs = [create_cap(grad,var) for grad, var in gvs]
+        capped_gvs = [x for x in capped_gvs if x != None]
+        return optimizer.apply_gradients(capped_gvs)
+
+
+    def build_optimizer(self, config, prefix, trainer_config, learning_rate, var_list, loss):
+        print("VARS Are", var_list)
```
Comment (martyn) [inline hypergan/trainers/base_trainer.py]: remove output

File: hypergan/trainers/alpha_trainer.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        for i, _ in enumerate(losses):
+            loss = losses[i]
+            var_list = self.var_lists[i]
+
+            if i ==0 or i == 1:
+                optimizer = self.build_optimizer(config, 'g_', config.g_trainer, self.g_lr, var_list, loss)
+            else:
+                optimizer = self.build_optimizer(config, 'd_', config.d_trainer, self.d_lr, var_list, loss)
+            optimizers.append(optimizer) #TODO prefx
+
```
Comment (martyn) [inline hypergan/trainers/alpha_trainer.py]: This is the second place where it would be nice to output a set subtraction of tf.global_variables and all the variables that have a gradient defined in any of the optimizers.  This would allow us to give warnings if a stray variable existed anywhere in the graph.
